spectator-ext-stackdriver/src/main/java/com/netflix/spectator/stackdriver/StackdriverWriter.java

```diff
@@ -0,0 +1,760 @@
+   *  This is a mechanism allowing external injection of these special cases.
+   *  The hints are only used when the need to create a descriptor arises.
+   *  Descriptors are only created once in the lifetime of a project (unless
+   *  the label is explicitly deleted, erasing all history, at some point).
+   *
+   *  We could externally create the labels, but doing so dynamically feels
+   *  more maintainable.
+   *
+   *  The key in this table is the desired Stackdriver descriptor name.
+   *
+   *  Note that this is not thread safe. Presumably we write to the hints
+   *  only on startup, and from within one thread so this doesnt matter.
+   */
+  private final Map<String, List<String>> labelHints
+      = new HashMap<String, List<String>>();
+
+  /**
+   * If Stackdriver were to return an error complaining about an TimeSeries
+   * data element we are trying to write, it references that by its index
+   * into the array we sent. That is pretty much useless to debug since we
+   * have no idea what or how the data is organized.
+   *
+   * This method is intended to interpret the Stackdriver error message
+   * in the context of the request we made and return the referenced element
+   * so that it can be logged in more explicit detail to understand what
+   * happend (most likely what label was missing from what descriptor type).
+   */
+  static public TimeSeries
+  findProblematicTimeSeriesElement(String msg, List<TimeSeries> nextN) {
+    String regex = "timeSeries\\[(\\d+?)\\]\\.metric\\.labels\\[(\\d+?)\\]";
+    Matcher matcher = Pattern.compile(regex).matcher(msg);
+    if (matcher.find()) {
+      int tsIndex = Integer.parseInt(matcher.group(1)) - 1;
+      TimeSeries ts = nextN.get(tsIndex);
+        return ts;
+    }
+    return null;
+  }
+
+  /**
+   * Factory method to create a writer.
+   *
+   * The customTypeNamespace is intended to act as a namespace for the
+   * stackdriver metrics found, such as the system name (e.g. spinnaker).
+   * This will help deconflict our metrics with another system that is
+   * also using Stackdriver Custom Metrics in the same project as us and
+   * might want to have CustomTypeDescriptors with the same name.
+   */
+  public static StackdriverWriter makeStackdriverWriter(
+      String customTypeNamespace,
+      String projectName,
+      String applicationName,
+      String applicationVersion,
+      String credentialsPath,
+      MeasurementFilter measurementFilter) {
+    Monitoring monitoring;
+
+    try {
+      HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
+      JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();
+      GoogleCredential credential = loadCredential(transport, jsonFactory,
+                                                   credentialsPath);
+      Monitoring.Builder builder = new Monitoring.Builder(
+            transport, jsonFactory, credential);
+
+        builder.setApplicationName("Spectator");
```

Comment:
It doesnt. This is deliberate but could have gone either way. I decided on Spectator because that is the component the user agent is acting on behalf of and I can later add the library version in as well (once I figure out how to do that). The application name is used in the reporting of where the actual metrics came from.

Application name is used to associate the data pushed into stackdriver but Spectator is used to identify the user agent interacting with stackdriver. I dont really need anything here, but it might be interesting to know in the future. A more accurate name would say something to the effect of spectator on behalf of the application name, but that level of precision isnt necessary.